server/src/main/java/org/elasticsearch/cluster/InternalClusterInfoService.java

```diff
@@ -358,13 +364,15 @@ public void onFailure(Exception e) {
         }
         ClusterInfo clusterInfo = getClusterInfo();
         boolean anyListeners = false;
-        for (final Consumer<ClusterInfo> listener : listeners) {
-            anyListeners = true;
-            try {
-                logger.trace("notifying [{}] of new cluster info", listener);
-                listener.accept(clusterInfo);
-            } catch (Exception e) {
-                logger.info(new ParameterizedMessage("failed to notify [{}] of new cluster info", listener), e);
+        synchronized (listeners) {
```

Comment:
It's always a little scary to invoke listeners under a lock IMO.
Can't we just use a `CopyOnWriteArrayList`  for the `listeners` field instead and be happy? (doesn't seem to matter for now anyway since there's only one spot in `Node` that ever calls this but still :))

Reply:
Sure, `CopyOnWriteArrayList` it is - d3f21ba94dd9496d0217c4ae6de0ba113ace724d